subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/NormalizedPathChangeDetector.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.history.changes;
+
+import com.google.common.collect.ListMultimap;
+import com.google.common.collect.MultimapBuilder;
+import org.gradle.internal.file.FileType;
+import org.gradle.internal.fingerprint.FileSystemLocationFingerprint;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.util.Map.Entry.comparingByKey;
+
+public class NormalizedPathChangeDetector implements CompareStrategy.ChangeDetector<FileSystemLocationFingerprint> {
+
+    public static final NormalizedPathChangeDetector INSTANCE = new NormalizedPathChangeDetector();
```

Comment:
Maybe NormalizedPathChangeDetector could be an enum.

Reply:
Enum has some nice properties, that are normally desired for a singleton: for example instantiation of enum is thread-safe by default, enum is final by default, it has private constructor by default, it's serializable by default.

From that, it simplifies writing a singleton, since you don't need to worry for example if instance is created in thread-safe manner.

Reply:
I also don't have problem with keeping the more explicit singleton pattern.

Reply:
I ended up converting this to an enum.